### Анализ результатов A/B-тестирования: доверительные интервалы

Итак, мы с вами знаем, что на основе выборки (части наблюдений) мы всегда можем измерить следующие показатели:
* выборочное среднее (центральная мера распределения);
* выборочное стандартное отклонение (мера разброса);
* выборочную пропорцию (доля объектов из выборки, выбранных по условию, например доля пользователей, совершивших покупки).

Однако эти оценки не являются нашим основным предметом интереса. На самом деле нас интересуют параметры генеральной совокупности: 
* истинное среднее — математическое ожидание ();
* истинное стандартное отклонение ();
* истинная пропорция ().

Если вы взяли выборку клиентов и выяснили их предпочтения, вам важно не то, что вы узнали предпочтения именно этих клиентов — вам важно, что теперь вы знаете предпочтения всех ваших клиентов (генеральная совокупность).

Важно помнить о разнице между выборочными статистиками и параметрами генеральной совокупности.

Выборочная статистика является случайной величиной, ведь она рассчитана на случайной выборке. Если мы возьмём одну выборку и посчитаем значение среднего, мы получим одно значение; возьмём другую — получим другое значение.

А вот параметр генеральной совокупности — это константа, конкретное число, которое, как правило, нам неизвестно и никогда известным не станет. Однако его можно приблизительно оценить!

Существует два способа оценки параметров генеральной совокупности:
⭐ Точечные оценки.
⭐ Интервальные оценки.

С точечными оценками на самом деле мы уже знакомы и не раз их вычисляли — это просто выборочные статистики. Например, мы собрали данные о росте прохожих людей и посчитали среднее, стандартное отклонение и так далее. Почему они точечные? Потому что они дают ответ точечно, то есть в виде одного числа.

С одной стороны, это удобно: взяли выборку, посчитали выборочное среднее и получили ответ, к примеру «средний рост в этой выборке равен 178 см». 

Однако есть у такого подхода и существенный недостаток: полученное значение практически никогда не равно истинному. 

**Интервальные оценки** — это ещё один способ оценки параметров генеральной совокупности, при использовании которого ответ даётся не в виде одного числа, а в виде интервала. 

Тогда мы скажем так: «Истинный средний рост лежит в интервале от 170 до 179 см с такой-то долей уверенности». Скорее всего, истинный параметр будет находиться внутри этого интервала.

Поиск интервальной оценки заключается в построении доверительного интервала. 

**Доверительный интервал** — интервал, который с заданной надёжностью покрывает значение неизвестного параметра. 

Например, если вам говорят, что построен 90 %-ый доверительный интервал для среднего роста и имеет вид: (170, 179), то это означает, что с надёжностью в 90 % истинный средний рост (его математическое ожидание) лежит в пределах от 170 см до 179 см. То есть с вероятностью в 90 % ожидается (в среднем), что любой человек из рассматриваемой генеральной совокупности имеет рост от 170 до 179 см.

Доверительные интервалы бывают двусторонними, левосторонними и правосторонними (как гипотезы). Мы с вами будем рассматривать только построение двусторонних доверительных интервалов, так как они являются наиболее популярными.

Любой двусторонний доверительный интервал обладает следующей структурой:

`Параметр = Выборочная оценка ± Предел погрешности`

Выборочной оценкой может быть среднее, стандартное отклонение, пропорция, медиана и так далее. А о том, как считается предел погрешности, и от чего он зависит, мы поговорим позже. 

В соответствии с этой формулой можно нарисовать доверительный интервал. Обозначим выборочную оценку как , а предел погрешности как . 

Тогда истинное значение параметра (с заданной степенью надёжности) лежит где-то внутри интервала:



На оси возможных значений параметра откладывается точка . Влево и вправо от  откладывается погрешность . Обратите внимание, что выборочная оценка всегда находится в центре двустороннего доверительного интервала. В результате построения интервала мы получаем зону, в которой предположительно лежит истинное значение параметра. 

#### ДОВЕРИТЕЛЬНЫЙ ИНТЕРВАЛ ДЛЯ СРЕДНЕГО ПРИ ИЗВЕСТНОМ СТАНДАРТНОМ ОТКЛОНЕНИИ

Самым частым параметром, который хотелось бы оценить, является истинное среднее (математическое ожидание). 

ОТКУДА БЕРЁТСЯ Z-КРИТИЧЕСКОЕ?

Вообще z в формуле доверительного интервала для среднего — это случайная величина, имеющая нормальное распределение с математическим ожиданием (центром) мю=0 и стандартным отклонением (разбросом) сигма=1 .

Компонент z-критическое в доверительном интервале определяется уровнем надёжности.

гамма=1-альфа, где альфа — уровень значимости или вероятность ошибки I рода. Тогда уровень надёжности — вероятность НЕ допустить ошибку I рода.

Вывод. Под уровнем надёжности понимается вероятность того, что истинное значение параметра окажется в построенном интервале. А под уровнем значимости — вероятность того, что построенный доверительный интервал «промахнётся» и не захватит истинное значение параметра.

То есть мы хотим взять такое значение, при котором в интервале от -z критического до z критического лежало бы 95 % всех исходов. В двустороннем интервале эти 95 % мы будем брать как раз посередине, чтобы справа и слева осталось 5 %, по 2.5 % с каждой стороны (помним, что нормальное распределение симметрично).

Таким образом, z критическое — это значение, которое отсекает критическую область нормального распределения. 

Найти значение z-критического для любого уровня значимости можно при помощи функции norm.ppf() из модуля scipy.stats. По специфическим причинам реализации функция возвращает не совсем z-критическое, а его отрицательный модуль, поэтому мы берём его со знаком минус.

In [3]:
from scipy.stats import norm

n = 36 # размер выборки
x_mean = 16100 # выборочное среднее
sigma = 12000 # истинное стандартное отклонение
gamma = 0.95 # уровень надёжности
alpha = 1 - gamma # уровень значимости

z_crit = -norm.ppf(alpha/2) # z критическое

eps = z_crit * sigma/(n ** 0.5) #погрешность
lower_bound = x_mean - eps # левая (нижняя) граница
upper_bound = x_mean + eps # правая (верхняя) граница
confidence_interval = (round(lower_bound), round(upper_bound)) # создаём кортеж из округлённых границ интервала
print('Доверительный интервал: {}'.format(confidence_interval)) # выводим результат

Доверительный интервал: (12180, 20020)


Примечание. Заметьте, что это универсальный код. Если перед вами возникнет другая задача расчёта доверительного интервала для истинного среднего, то вам достаточно будет правильно поменять значения переменных. Вы можете оформить его в виде функции и сохранить её, чтобы не строить каждый раз новый доверительный интервал

Полученный результат интерпретируется следующим образом: мы на 95 % уверены, что истинные расходы жителей на продукты питания лежат в интервале от 12 180 рублей до 20 020 рублей. 

⚡ Важно! Доверительный интервал не даёт нам 100 % результата. Всегда есть шанс, что мы промахнулись и не захватили истинного значения. Этот шанс как раз равен уровню значимости .
То есть в нашем примере существует 5 %-ый шанс, что истинное среднее расходов вовсе не лежит в интервале от 12 180 до 20 020. Существует вероятность, что математическое ожидание на самом деле равно 5 000 рублей или 25 000 рублей. Но такая вероятность составляет менее 5 %, и это нас устраивает.

#### ДОВЕРИТЕЛЬНЫЙ ИНТЕРВАЛ ДЛЯ СРЕДНЕГО ПРИ НЕИЗВЕСТНОМ СТАНДАРТНОМ ОТКЛОНЕНИИ

Для правомерного использования формулы истинное стандартное отклонение выборки должно быть известно. Но, как правило, нам неизвестно это значение (если только оно не свалилось на нас озарением), ведь для этого нам нужно знать все значения генеральной совокупности. А если они у нас есть, то нам и доверительные интервалы не нужны — зная всю генеральную совокупность, всегда можно найти истинные параметры.

В реальной жизни нам приходится заменять значение  значением выборочного стандартного отклонения, которое мы можем посчитать на основании выборки.

Когда мы заменяем константу на случайную величину, точность всей конструкции явно становится меньше. Вероятно, итоговая величина будет похожа на нормальное распределение, но будет обладать бОльшим разбросом.

Такая случайная величина хорошо описывается распределением Стьюдента, или Т-распределением. 

Распределение Стьюдента — это целое семейство родственников стандартного нормального распределения.

Действительно, центр (среднее) по-прежнему будет в 0, а вот разброс значений (то есть стандартное отклонение) будет уже больше. Этот разброс будет различным в зависимости от степеней свободы (k).

Степени свободы  обозначаются d.f. (от англ. degrees of freedom). Они и говорят нам о том, насколько велик разброс в значениях распределения Стьюдента. Этот параметр вычисляется как , где  — размер выборки.

Найти значение t-критического для любого уровня значимости можно, используя функцию t.ppf() из модуля scipy.stats. Для этого в функцию передаётся интересующий уровень значимости () и число степеней свободы (). Аналогично z-критическому функция возвращает отрицательный модуль t-критического, поэтому мы берём его со знаком минус:

In [7]:
from scipy.stats import t

n = 15 # размер выборки
k = n - 1 # число степеней свободы
x_mean = 2000 # выборочное среднее
x_std = 400 # выборочное стандартное отклонение
gamma = 0.95 # уровень надёжности
alpha = 1 - gamma # уровень значимости

t_crit = -t.ppf(alpha/2, k) # t-критическое

eps = t_crit * x_std/(n ** 0.5) # погрешность
lower_bound = x_mean - eps # левая (нижняя) граница
upper_bound = x_mean + eps # правая (верхняя) граница
confidence_interval = (round(lower_bound), round(upper_bound)) # создаём кортеж из округлённых границ интервала
print('Доверительный интервал: {}'.format(confidence_interval)) # выводим результат

Доверительный интервал: (1778, 2222)


Примечание. Важно отметить, что при достаточной выборке (примерно более 30 наблюдений) распределение Стьюдента очень похоже на стандартное нормальное распределение. В этом случае можно использовать первую формулу (z-критическое) даже при неизвестном стандартном отклонении. Однако используя вторую формулу при оценке среднего, вы не ошибётесь.

Ещё один важный факт состоит в том, что обе формулы работают только при соблюдении двух условий:
1. Выборка должна быть случайной.
2. Выборка должна быть достаточно велика. Но есть исключение: если выборка маленькая, но сама величина распределена нормально, то выборочное среднее тоже будет распределено нормально. В таком случае мы можем пользоваться формулой доверительного интервала.

In [10]:
n = 25 # размер выборки
k = n - 1 # число степеней свободы
x_mean = 3540 # выборочное среднее
x_std = 1150 # выборочное стандартное отклонение
gamma = 0.9 # уровень надёжности
alpha = 1 - gamma # уровень значимости

t_crit = -t.ppf(alpha/2, k) # t-критическое
print(t_crit)

eps = t_crit * x_std/(n ** 0.5) # погрешность
lower_bound = x_mean - eps # левая (нижняя) граница
upper_bound = x_mean + eps # правая (верхняя) граница
confidence_interval = (round(lower_bound, 2), round(upper_bound, 2)) # создаём кортеж из округлённых границ интервала
print('Доверительный интервал: {}'.format(confidence_interval)) # выводим результат

1.7108820799094284
Доверительный интервал: (3146.5, 3933.5)


#### ДОВЕРИТЕЛЬНЫЙ ИНТЕРВАЛ ДЛЯ ПРОПОРЦИИ

Настало время вернуться к нашей задаче об анализе конверсии в двух группах A/B-тестирования.

Конверсия — это доля пользователей, совершивших целевое действие, по отношению к общей массе пользователей. 

Ранее мы уже выяснили, что признак того, осуществлена конверсия пользователем или нет — это случайная величина, имеющая распределение Бернулли. Такая случайная величина имеет два возможных исхода: успех (1) и неуспех (0).

То есть конверсия — это вероятность того, что пользователь совершит целевое действие (вероятность успеха традиционно обозначается как ):
количество успехов (количество пользователей, осуществивших конверсию)/общее количество исходов (количество пользователей).

Примечание. При расчётах конверсии обычно долю ещё умножают на 100 %.

В контексте задачи A/B-тестирования нам бы хотелось знать истинную конверсию (а не ту, которая подсчитана по выборке) в контрольной и тестовой группе, а также оценить истинную разницу между конверсиями. 

Если подумать, то на самом деле мы решаем задачу оценки истинной пропорции генеральной совокупности по выборке. А значит,  нужно научиться строить доверительный интервал для пропорции.

Забегая вперёд в раздел по теории вероятностей, отметим, что истинное среднее (математическое ожидание) такой случайной величины , а истинное стандартное отклонение . 

Так как истинную вероятность успеха , мы, конечно же, не знаем, так как нам неизвестна вся генеральная совокупность (всех возможных пользователей), то вместо неё мы используем выборочную пропорцию (обозначим её как ) — долю совершивших конверсию пользователей, то есть выборочную конверсию. 

Заметим, что истинная пропорция  полностью совпадает с истинным средним . Если немного поколдовать с формулой доверительного интервала для истинного среднего, то можно вывести формулу доверительного интервала для истинной пропорции (истинной конверсии):

In [11]:
import pandas as pd

ab_data = pd.read_csv('ab_data.csv')
ab_data.head()

user_id   timestamp group  converted
0   851104  2017-01-21     A          0
1   804228  2017-01-12     A          0
2   661590  2017-01-11     B          0
3   853541  2017-01-08     B          0
4   864975  2017-01-21     A          1

In [12]:
#фильтруем данные группы А
a_data = ab_data[ab_data['group'] == 'A']
#фильтруем данные группы B
b_data = ab_data[ab_data['group'] == 'B']

Давайте на этот раз оформим расчёт в виде функции: объявим функцию proportions_conf_interval(). У данной функции будет три аргумента: n — размер выборки, x_p — выборочная пропорция и gamma — уровень надёжности (по умолчанию он равен 0.95). Функция будет возвращать кортеж из вычисленных границ доверительного интервала, умноженных на 100 % и округлённых до второго знака после запятой.

In [14]:
def proportions_conf_interval(n, x_p, gamma=0.95):   
    alpha = 1 - gamma # уровень значимости
    z_crit = -norm.ppf(alpha/2) # z критическое
    eps = z_crit * (x_p * (1 - x_p) / n) ** 0.5 #погрешность
    lower_bound = x_p - eps # левая (нижняя) граница
    upper_bound = x_p + eps # правая (верхняя) граница
    # возвращаем кортеж из округлённых границ интервала
    return round(lower_bound * 100, 2), round(upper_bound * 100, 2)

conf_interval_a = proportions_conf_interval(
n=a_data['user_id'].count(), # размер выборки
x_p=a_data['converted'].mean() # выборочная пропорция
)
conf_interval_b = proportions_conf_interval(
n=b_data['user_id'].count(), # размер выборки
x_p=b_data['converted'].mean() # выборочная пропорция
)
print('Доверительный интервал для конверсии группы А: {}'.format(conf_interval_a))
print('Доверительный интервал для конверсии группы B: {}'.format(conf_interval_b))

Доверительный интервал для конверсии группы А: (11.86, 12.19)
Доверительный интервал для конверсии группы B: (11.7, 12.03)


Итак, можно сделать вывод, что с надёжностью в 95 % конверсия группы А лежит в интервале от 11.86 % до 12.19 %, а конверсия группы B лежит в интервале от 11.7 % до 12.03 %.

Что значит пересечение интервалов для нас? Это значит, что существует вероятность, что истинное значение конверсии в обеих группах может совпасть. То есть мы не можем однозначно сказать, есть ли разница в конверсиях. Можно даже рассчитать эту вероятность, но лучше пойти другим путём.

Лучше построить доверительный интервал разницы пропорций (конверсий).

Не будем уходить в статистические дебри, просто приведём формулу для расчёта доверительного интервала разницы конверсий в группах. Запоминать её, как и считать руками, не нужно, мы реализуем расчёты с помощью Python. Однако важно понимать её составляющие:

✏️ Создадим функцию diff_proportions_conf_interval() со следующими аргументами: n — список из размеров выборки для групп А и B соответственно, xp — список из выборочных пропорций для групп А и B соответственно и gamma — уровень надёжности (по умолчанию он равен 0.95). Функция будет возвращать кортеж из вычисленных границ доверительного интервала разницы конверсий, умноженных на 100 % и округлённых до второго знака после запятой.

In [24]:
def diff_proportions_conf_interval(n, xp, gamma=0.95):
    alpha = 1 - gamma # уровень значимости
    diff = xp[1] - xp[0] # выборочная разница конверсий групп B и A
    z_crit = -norm.ppf(alpha/2) # z критическое
    eps = z_crit * (xp[0] * (1 - xp[0])/n[0] + xp[1] * (1 - xp[1])/n[1]) ** 0.5 # погрешность
    lower_bound = diff - eps # левая (нижняя) граница
    upper_bound = diff + eps # правая (верхняя) граница
    # возвращаем кортеж из округлённых границ интервала
    return round(lower_bound *100, 2), round(upper_bound * 100, 2)

# размеры выборок групп А и B
n = [a_data['user_id'].count(), b_data['user_id'].count()]
# выборочная пропорция групп A и B
xp = [a_data['converted'].mean(), b_data['converted'].mean()]
# строим доверительный интервал для разности конверсий
diff_inverval = diff_proportions_conf_interval(n, xp)
print('Доверительный интервал для разности конверсий: {}'.format(diff_inverval))

Доверительный интервал для разности конверсий: (-0.39, 0.08)


Итак, что можно сказать теперь? Построенный доверительный интервал с 95 %-ой надёжностью говорит нам, что разница в конверсии между вариантом B и вариантом А лежит в интервале от -0.39 % до 0.08 %. 

То есть разница в конверсии с 95 %-ой надёжностью может быть как положительной, так и отрицательной. Мы имеем дело в третьим случаем — интервал охватывает точку 0.

Отсюда можно сделать вывод, что статистической разницы в конверсии между вариантами А и B не существует, то есть они являются равнозначными по данной метрике.

В заключение хотелось бы отметить, что доверительный интервал для пропорции применяется не только для оценки конверсии и разницы конверсий в группах!

Под истинной пропорцией может пониматься доля успешно угаданных моделью объектов (accuracy), доля выздоровевших пациентов или доля ушедших клиентов (отток). То есть такой доверительный интервал может быть применён в любой задаче A/B-тестирования, где признак, по которому считается метрика, бинарный (имеет два возможных исхода — «успех» и «неуспех»). А это очень большой спектр задач!

Таким образом, доверительные интервалы — это ещё один вспомогательный инструмент проведения A/B-тестирования, который позволяет с определённой заранее надёжностью статистически сравнить варианты продукта А и B по заданной метрике. 

⭐️ Обычно доверительные интервалы используются в комплексе со статистическими тестами. Статистический тест показывает, значимы ли наши результаты, а доверительный интервал показывает пределы изменения метрик и их разницы в группах. 

In [20]:
def proportions_conf_interval(n, x_p, gamma=0.95):   
    alpha = 1 - gamma # уровень значимости
    z_crit = -norm.ppf(alpha/2) # z критическое
    eps = z_crit * (x_p * (1 - x_p) / n) ** 0.5 #погрешность
    lower_bound = x_p - eps # левая (нижняя) граница
    upper_bound = x_p + eps # правая (верхняя) граница
    # возвращаем кортеж из округлённых границ интервала
    return round(lower_bound * 100, 2), round(upper_bound * 100, 2)

conf_interval_a = proportions_conf_interval(
n=100, # размер выборки
x_p=45/1000 # выборочная пропорция
)
conf_interval_b = proportions_conf_interval(
n=1000, # размер выборки
x_p=50/1000 # выборочная пропорция
)
print('Доверительный интервал для конверсии группы А: {}'.format(conf_interval_a))
print('Доверительный интервал для конверсии группы B: {}'.format(conf_interval_b))

Доверительный интервал для конверсии группы А: (0.44, 8.56)
Доверительный интервал для конверсии группы B: (3.65, 6.35)


#### Закрепление знаний

→ Прежде чем мы перейдём к выполнению финального задания, давайте ещё раз обобщим всё, что мы изучили в модуле.

* Итак, A/B-тестирование проводится для того, чтобы сравнить два (или более) варианта продукта по их эффективности (значениям метрик).
* Процесс заключается в разделении аудитории на группы и предоставлении каждой группе своего варианта продукта. На протяжении тестирования собираются данные о метриках, и на основании результатов делается вывод, какой из вариантов лучше.
* A/B-тесты — очень важный инструмент, особенно если ваш продукт основан на взаимодействии с пользователями. Если вы хотите протестировать новую идею, но не знаете, как оценить результат, то A/B-тестирование станет для вас верным другом. Оно позволяет узнать предпочтения ваших пользователей и понять, какие изменения в продукт стоит вводить.
* На каждом из этапов формулируются свои результаты, которые затем используются на следующих этапах. Главная цель A/B-тестирования — принять взвешенное, статистически доказанное решение по продукту.
* Помним, что при проведении A/B-теста нужно следовать трём принципам:
1. исключать влияние извне;
2. использовать большое количество данных;
3. применять правильные инструменты для анализа.
* Мы с вами сконцентрировались на этапе анализа результатов и формировании выводов. Для того чтобы грамотно оценить результаты A/B-теста, нам нужно выполнить следующие шаги:
1. подготовить наши данные к анализу и убедиться, что соблюдены принципы A/B-тестирования (в каждой группе собрано достаточно данных и длительность теста одинакова, группы сбалансированы и отсутствуют пересечения в группах);
2. визуально оценить интересующие нас метрики в каждой из групп, чтобы сделать первичные выводы;
3. убедиться в том, что метрики стабилизированы и мы можем проводить статистические тесты (помним, о проблеме подглядывания). Для этого чаще всего используются кумулятивные метрики (метрика считается стабилизированной во времени, если её кумулятивный вариант переходит в прямую на графике);
4. если метрика стабилизировалась, то мы можем проводить статистические тесты, которые позволяют проверить гипотезу о равенстве показателей в группах при альтернативных гипотезах (неравенство/больше/меньше), но предварительно необходимо узнать распределение показателя, который мы пытаемся сравнить (помним о том, что каждый тест имеет свою область применения);
5. в комплексе со статистическими тестами мы можем использовать доверительные интервалы для параметра, чтобы оценить его возможные границы;
на основе результатов статистических тестов и построенных доверительных интервалов мы формируем вывод и принимаем решение.